Classes/ContentRepository/NodeTranslationService.php

```diff
@@ -210,39 +191,129 @@ protected function translateNode(NodeInterface $sourceNode, NodeInterface $targe
 
         if (count($propertiesToTranslate) > 0) {
             $translatedProperties = $this->translationService->translate($propertiesToTranslate, $targetLanguage, $sourceLanguage);
-            $translatedProperties = array_merge($translatedProperties, $properties);
-            foreach ($translatedProperties as $propertyName => $translatedValue) {
-                if ($targetNode->getProperty($propertyName) != $translatedValue) {
-                    $targetNode->setProperty($propertyName, $translatedValue);
-                }
+            $properties = array_merge($translatedProperties, $properties);
+        }
+
+        foreach ($properties as $propertyName => $propertyValue) {
+            if ($targetNode->getProperty($propertyName) != $propertyValue) {
+                $targetNode->setProperty($propertyName, $propertyValue);
             }
         }
     }
 
     /**
-     * @param  string  $language
+     * @param  string  $targetLanguage
+     * @param  string|null  $sourceLanguage
      * @param  string  $workspaceName
      * @return Context
      */
-    protected function getContextForLanguageDimensionAndWorkspaceName(string $language, string $workspaceName = 'live'): Context
+    public function getContextForTargetLanguageDimensionAndSourceLanguageDimensionAndWorkspaceName(string $targetLanguage, string $sourceLanguage = null, string $workspaceName = 'live'): Context
     {
-        $dimensionAndWorkspaceIdentifierHash = md5($language . $workspaceName);
+        $dimensionAndWorkspaceIdentifierHash = md5(trim($sourceLanguage . $targetLanguage . $workspaceName));
 
         if (array_key_exists($dimensionAndWorkspaceIdentifierHash, $this->contextFirstLevelCache)) {
             return $this->contextFirstLevelCache[$dimensionAndWorkspaceIdentifierHash];
         }
 
+        $languageDimensions = array($targetLanguage);
+        if (!is_null($sourceLanguage)) {
+            $languageDimensions[] = $sourceLanguage;
+        }
```

Comment:
Is that safe …afaik this creates a fallback from the source to the target that is unlikely to be intended

Reply:
```
dimensionConfig->$languageDimension->presets->$targetLanguage->values